app/controllers/application_controller.rb

```diff
@@ -5,6 +5,151 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    set :sessions, true
+    set :method_override, true
+    set :session_secret, "fwitter_efrain"
   end
 
-end
\ No newline at end of file
+  get "/" do
+    erb :index
+  end
+
+  get "/signup" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :"users/create"
+    end
+  end
+
+  post "/signup" do
+
+    unless params[:username] == "" || params[:password] == "" || params[:email] == ""
+      User.create(username: params[:username], password: params[:password], email: params[:email])
+      session[:id] = User.last.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @tweets = all_tweets
+      erb :"tweets/tweets"
+    else
+      redirect '/login'
+    end
+  end
+
+  get "/tweets/new" do
+    if logged_in?
+      erb :"tweets/create"
+    else
+      redirect "/login"
+    end
+  end
+
+  post "/tweets" do
+    unless params[:content] == ""
+      current_user.tweets << Tweet.create(content: params[:content])
+      redirect "/tweets"
+    else
+      redirect "/tweets/new"
+    end
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show'
+    else
+      redirect "/login"
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == session[:id]
+        erb :'tweets/edit'
+      else
+        redirect "/tweets"
+      end
+    else
+      redirect "/login"
+    end
+  end
+
+  patch "/tweets/:id" do
+    unless params[:content] == ""
+      Tweet.update(params[:id], content: params[:content])
+      redirect "/tweets"
+    else
+      redirect "/tweets/#{params[:id]}/edit"
+    end
+  end
+
+  get "/login" do
+    unless logged_in?
+      erb :'users/login'
+    else
+      redirect '/tweets'
+    end
+
+  end
+
+  post "/login" do
+    unless params[:username] == "" || params[:password] == ""
+      user = User.find_by(username: params[:username])
+      if !!user && !!user.authenticate(params[:password])
+        session[:id] = user.id
+        redirect '/tweets'
+      else
+        redirect '/login'
+      end
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.delete(:id)
+      redirect "/login"
+    else
+      redirect "/"
+    end
+  end
+
+  get "/users/:slug" do
+    @tweets = all_tweets_by_user(params[:slug])
+    erb :"tweets/tweets"
+  end
+
+  delete "/tweets/:id" do
+    tweet = Tweet.find_by_id(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.